dbms/src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -855,7 +855,35 @@ void SelectQueryExpressionAnalyzer::appendProjectResult(ExpressionActionsChain &
         String result_name = ast->getAliasOrColumnName();
         if (required_result_columns.empty() || required_result_columns.count(result_name))
         {
-            result_columns.emplace_back(ast->getColumnName(), result_name);
+            std::string source_name = ast->getColumnName();
+
+            /*
+             * For temporary columns created by ExpressionAnalyzer for literals,
+             * use the correct source column. Using the default display name
+             * returned by getColumnName is not enough, and we have to use the
+             * column id set by EA. In principle, this logic applies to all kinds
+             * of columns, not only literals. Literals are especially problematic
+             * for two reasons:
+             * 1) confusing different literal columns leads to weird side
+             *    effects (see 01101_literal_columns_clash);
+             * 2) the disambiguation mechanism in SyntaxAnalyzer, that, among
+             *    other things, creates unique aliases for columns with same
+             *    names from different tables, is applied before these temporary
+             *    columns are created by ExpressionAnalyzer.
+             * Similar problems should also manifest for function columns, which
+             * are likewise created at a later stage by EA.
+             * In general, we need to have explicit separation between display
+             * names and identifiers for columns. This code is a workaround for
+             * a particular subclass of problems, and not a proper solution.
+             */
+            if (auto as_literal = dynamic_cast<ASTLiteral *>(ast.get());
+                as_literal)
+            {
+                source_name = as_literal->unique_column_name;
```

Comment:
I'd prefer to make `unique_column_name` optional and move it to `ASTWithAlias` or even `IAST`. Then we allow to reuse it in general case described in comment. So we could make comment shorter :)
The logic is: if we set it we need use it instead of source_name. It's self descriptive.

Reply:
I have mixed feelings about this one. We only apply this hack for literals, but having this field for all aliases would give an false impression that it also applies to some other kinds of columns, when in fact it does not. If I added a proper display name/column id separation in this PR, then sure, it would be be better to make this general. But it's not general now, and we better make this clear. Maybe I should rename this column to `unique_literal_column_name` instead.